### PR TITLE
Translation Integration with Transifex - Phase 1

### PR DESCRIPTION
**Pull request by mkangia**
This is the first phase of integration.

This adds the ability to send files to transifex for translation for a particular build of an app.
Heads up that the implementation for this might add things when follow up PR is done, for ex: add ability to send translation files with translations for target languages as well or add better occurrences for the strings.

A follow up PR would address pulling files from transifex and then adding those translations

### CODE REVIEW

File: custom/icds/translations/integrations/client.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from __future__ import absolute_import
+import requests
+import os
+import json
+
+
+class TransifexApiClient():
+    def __init__(self, token, orgranization, project):
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: eye of the tiger, thanks 👍

File: custom/icds/translations/integrations/client.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        return requests.get(url, auth=self._auth)
+
+    def delete_resource(self, resource):
+        url = "https://www.transifex.com/api/2/project/{}/resource/{}".format(
+            self.project, resource)
+        return requests.delete(url, auth=self._auth)
+
+    def upload_resource(self, path_to_pofile, resource_slug, resource_name):
+        url = "https://www.transifex.com/api/2/project/{}/resources".format(self.project)
+        content = open(path_to_pofile, 'r').read()
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: ya, i don't think it will retain keep any reference to the file.

> is the pofile kept as a temporary file on the filesystem?
No its not. Its generated in the current directory and then deleted at the end of the process here itself. I would like us to remove them here because they can be many in number and in case this rolls out as a feature for multiple domains and is used, it would result in quite some files which would stay around till they are removed automatically.
Or can also make them temp file but still remove them if that sounds better, though not sure why, possibly in case this process fails and does not reach the step to remove them manually?

File: custom/icds/translations/integrations/transifex.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+from collections import namedtuple, OrderedDict
+from corehq.apps.app_manager import id_strings
+from custom.icds.translations.integrations.client import TransifexApiClient
+
+Translation = namedtuple('Translation', 'key translation occurrences')
+Unique_ID = namedtuple('UniqueID', 'type id')
+
+
+class Transifex():
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: I think having `Transifex` still makes sense to handle what needs to be sent and then processing whats received whereas `TransifexClient` just handles sending/pulling from transifex.
Agree about extracting the functionality to generate po files. will do that.

File: custom/icds/translations/integrations/transifex.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        {
+            sheet_name_with_build_id: {
+                key: Translation(key, translation, occurrences)
+            }
+        }
+        """
+        # get app or the app for the build
+        self.find_build_id()
+        from corehq.apps.app_manager.dbaccessors import get_app
+        app = get_app(self.domain, self.app_id_to_build)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: haha.. ohkay!

File: custom/icds/translations/integrations/transifex.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        from corehq.apps.app_manager.app_translations.app_translations import expected_bulk_app_sheet_rows
+        # simply the rows of data per sheet name
+        rows = expected_bulk_app_sheet_rows(app)
+
+        # get the translation data headers
+        from corehq.apps.app_manager.app_translations.app_translations import expected_bulk_app_sheet_headers
+        for header_row in expected_bulk_app_sheet_headers(app):
+            self.headers[header_row[0]] = header_row[1]
+        self.set_sheet_name_to_module_or_form_mapping(rows[u'Modules_and_forms'])
+        return rows
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: this might be tricky. Setting headers and that mapping relies to have the `rows` data populated. so its like a series of steps that happen. Let me try to improve this while i move this out to a `PoFilesUtility`.

File: custom/icds/translations/integrations/transifex.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        self.source_lang = source_lang
+        self.target_lang = target_lang
+        self.lang_prefix = lang_prefix
+        if not self.target_lang:
+            self.target_lang = self.source_lang
+        self.app_id_to_build = None
+        self.translations = OrderedDict()
+        self.headers = dict()  # headers for each sheet name
+        self.sheet_name_to_module_or_form_type_and_id = dict()
+        self.generated_files = list()
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: what do you mean by a single resource here?

File: custom/icds/translations/integrations/transifex.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        unique_id_column_index = self.get_header_index(u'Modules_and_forms', 'unique_id')
+        type_column_index = self.get_header_index(u'Modules_and_forms', 'Type')
+        for row in all_module_and_form_details:
+            self.sheet_name_to_module_or_form_type_and_id[row[sheet_name_column_index]] = Unique_ID(
+                row[type_column_index],
+                row[unique_id_column_index]
+            )
+
+    def get_filename(self, sheet_name):
+        return sheet_name + '_v' + str(self.version)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: Replied to a similar thing above https://github.com/dimagi/commcare-hq/pull/20492#discussion_r187567030
good point about conflicts. Can add a timestamp to it to avoid conflict and also a quick view of when the file was created, though that timestamp exists in the file itself as well.
Ya not needed across requests.

File: custom/icds/tasks.py [hunk 8]
```diff
@@ -44,3 +45,11 @@ def _get_query(db_name, max_age=max_age):
...
         if run_again:
             delete_old_images.delay()
+
+
+@task
+def send_translation_files_to_transifex(domain, form_data):
+    Transifex(domain, form_data.get('app_id'),
+              form_data.get('source_lang'),
+              form_data.get('target_lang'),
+              form_data.get('version')).send_files_to_transifex()
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/tasks.py]: `send_files_to_transifex` is the single method call. It does all steps internally then.

File: custom/icds_reports/forms.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class AppTranslationsForm(forms.Form):
+    app_id = forms.ChoiceField(label=ugettext_noop("App"), choices=(), required=True)
+    version = forms.IntegerField(label=ugettext_noop("Build Number"), required=True)
+    source_lang = forms.ChoiceField(label=ugettext_noop("Source Language"),
+                                    choices=[('en', 'English')])
+    target_lang = forms.ChoiceField(label=ugettext_noop("Source Language"),
+                                    choices=[('hin', 'Hindi'), ('mr', 'Marathi'),
+                                             ('te', 'telugu')])
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds_reports/forms.py]: Good question. Have not thought of that but if it is to be done eventually, would not be app config but whats defined on transifex as the supported source languages. For now just added options expected for icds.

File: corehq/apps/app_manager/dbaccessors.py [hunk 10]
```diff
@@ -431,11 +431,10 @@ def get_all_built_app_results(domain, app_id=None):
...
     from .models import Application
-    result = Application.get_db().view(
-        'app_manager/applications_brief',
-        startkey=[domain, app_id],
-        endkey=[domain, app_id, {}]
-    )
+    result = Application.get_db().view('app_manager/saved_app',
+                                       startkey=[domain, app_id, {}],
+                                       endkey=[domain, app_id],
+                                       descending=True)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/dbaccessors.py]: looks like thats happening already, don't see docs included till i pass `include_docs=True`. Not sure why though

File: custom/icds/translations/integrations/transifex.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply custom/icds/translations/integrations/transifex.py]: I mean one sheet/file - one set of translations, probably corresponding to a single form or module.

Right now, `translations`, `headers`, `sheet_name_to_module_or_form_type_and_id`, and `generated_files` each store a collection of data broken up by file.  For example, `self.translations` is a dict mapping file names to the translations for that file.

Instead, you could have a single property called `self.resources` (or whatever), which is a list of `Resource` objects (or a map of filenames to `Resource` objects).  Then each `Resource` object would manage it's own translations, headers, module/form type and id, and so forth.

This way, the broad strokes would happen in this parent class, and the child `Resource` class instances would handle the detail work, but they'd have a smaller set of state - only that corresponding to a single form or module.

File: corehq/apps/app_manager/dbaccessors.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/dbaccessors.py]: oh yeah, I think couchdbkit figures out which to use as the default for `include_docs` and `reduce` based on some set of parameters.  I can never remember which one is which, so I always specify if there's any confusion.

**Review by esoergel — APPROVED** [review]
thanks for the follow-up, looks good!

File: custom/icds/translations/integrations/transifex.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: hmm, all the translations and headers are received at once and `sheet_name_to_module_or_form_type_and_id` would also be generated at once from the first sheet. Its then essentially this method `_get_translation_for_sheet` that generates translation for a resource as needed for a po file, so it would be essentially be just this method in `Resource`. Will try that out to see if that makes sense to extract out.

**Comment by mkangia** [comment]
Thanks for the close reviews @esoergel . Was working on one more fix from last review, extracted POFilegenerator.

**Comment by mkangia** [comment]
@esoergel let me know if you have more inputs on this, if g2g going to merge this now.

File: settings.py [hunk 11]
```diff
@@ -925,7 +925,7 @@
 DAYS_TO_KEEP_DEVICE_LOGS = 60
 
 MAX_RULE_UPDATES_IN_ONE_RUN = 10000
-
+TRANSIFEX_DETAILS = None
```
[the remark this answers is not in the document]
Comment (mkangia) [reply settings.py]: Added a todo for this

File: custom/icds/translations/integrations/client.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        self.token = token
+        self.organization = orgranization
+        self.project = project
+
+    @property
+    def _auth(self):
+        return (self.username, self.token)
+
+    def list_resources(self):
+        url = "http://api.transifex.com/organizations/{}/projects/{}/resources".format(
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: tried that but looks like thats not supported. Got these links from their docs.

File: corehq/apps/app_manager/app_translations/generators.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+        translations_for_sheet = []
+        key_lang_index = self._get_header_index(sheet_name, self.lang_prefix + self.key_lang)
+        source_lang_index = self._get_header_index(sheet_name, self.lang_prefix + self.source_lang)
+        if sheet_name == MODULES_AND_FORMS_SHEET_NAME:
+            type_index = self._get_header_index(MODULES_AND_FORMS_SHEET_NAME, 'Type')
+            sheet_name_index = self._get_header_index(MODULES_AND_FORMS_SHEET_NAME, 'sheet_name')
+            unique_id_index = self._get_header_index(MODULES_AND_FORMS_SHEET_NAME, 'unique_id')
+
+            def occurrence(_row):
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/generators.py]: Not sure what to do here. Open to suggestions.

File: corehq/apps/app_manager/app_translations/generators.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        self.domain = domain
+        self.app_id = app_id
+        self.key_lang = key_lang
+        self.source_lang = source_lang
+        self.lang_prefix = lang_prefix
+        self.translations = OrderedDict()
+        self.version = version
+        self.headers = dict()  # headers for each sheet name
+        self.generated_files = list()  # list of tuples (filename, filepath)
+        self.sheet_name_to_module_or_form_type_and_id = dict()
```
Comment (esoergel) [inline corehq/apps/app_manager/app_translations/generators.py]: I do find all this business around sheets, rows, and column indices pretty confusing.  It's a little hard to follow, particularly since the application and transifex don't have anything to do with spreadsheets - that's just how we happen to have easy access to it, since that already exists.

You can punt on this if you don't think it's worth the effort, but you could modify `expected_bulk_app_sheet_rows` to return a custom data model that's easier to work with.

**Review by esoergel — APPROVED** [review]

File: custom/icds/tasks.py [hunk 15]
```diff
@@ -49,8 +49,16 @@ def _get_query(db_name, max_age=max_age):
...
+                  data.get('source_lang'),
+                  data.get('transifex_project_slug'),
+                  data.get('version')).send_translation_files()
+    if data.get('target_lang'):
+        Transifex(domain,
+                  data.get('app_id'),
+                  data.get('target_lang'),
+                  data.get('transifex_project_slug'),
+                  data.get('version'),
+                  is_source_file=False).send_translation_files()
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/tasks.py]: Yup, fixed later.

File: custom/icds/translations/integrations/client.py [hunk 16]
```diff
@@ -44,6 +44,22 @@ def upload_resource(self, path_to_pofile, resource_slug, resource_name):
...
+        if resource_name is None:
+            __, filename = os.path.split(path_to_pofile)
+            resource_name = filename
+        headers = {'content-type': 'application/json'}
+        data = {
+            'name': resource_name, 'slug': resource_slug, 'content': content,
+            'i18n_type': 'PO'
+        }
+        return requests.put(
+            url, data=json.dumps(data), auth=self._auth, headers=headers,
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: looks like that causing some issue when sending the request, possibly due to file content. So leaving this as-is.

File: corehq/apps/app_manager/app_translations/generators.py (hunk 14 above)
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/generators.py]: Agree.. but going to skip this for now and try to sneak this in during improvements later.

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]

File: corehq/apps/app_manager/app_translations/generators.py [hunk 17]
```diff
@@ -96,7 +96,10 @@ def occurrence(_row):
                 list_or_detail_index = self._get_header_index(sheet_name, 'list_or_detail')
 
                 def occurrence(_row):
-                    return ':'.join([_row[case_property_index], _row[list_or_detail_index]])
+                    case_property = _row[case_property_index]
+                    # limit case property length to avoid errors at Transifex where there is a limit of 1000
+                    case_property = case_property[:950]
+                    return ':'.join([case_property, _row[list_or_detail_index]])
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/generators.py]: Can you share an example @esoergel . Sorry i didn't understand that. Are you referring to the length limit i added? If yes, that is to address very long case property calculations, i don't think anything else would be that long enough.

[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/app_translations/generators.py]: Yeah, I was thinking of the case_property truncation.  What do you mean "case property calculations"?  I was just thinking if there _are_ user-defined property names that are 1000 characters long, it's conceivable that the first 950 characters of two properties might be the same.  

Users frequently use patterns like this in their property names:
`patient_2_demographic_information_age`
`patient_2_demographic_information_sex`
But at any rate, it does seem pretty unlikely that this would happen at that length.

[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/generators.py]: we found this long if expressions in excel sheet that was more than 1000 characters. Didn't come across a property name that was that long.. definitely highly unlikely someone would keep a name that long.. also looks like we wont rely on that to be unique. Might get more insight when i begin working on the second phase.

**Comment by mkangia** [comment]
Going to merge this now... if there is still something pending can take it up in a following PR.
